File: typescript/bc-lifehash/src/color-grid.ts
import { Color } from './color.js';
import { type ColorFunc } from './color-func.js';
import { FracGrid } from './frac-grid.js';
import { Grid } from './grid.js';
import { Pattern } from './patterns.js';

interface Transform {
    transpose: boolean;
    reflectX: boolean;
    reflectY: boolean;
}

export class ColorGrid {
    readonly grid: Grid<Color>;

    constructor(fracGrid: FracGrid, gradient: ColorFunc, pattern: Pattern) {
        const multiplier = pattern === Pattern.Fiducial ? 1 : 2;
        const targetWidth = fracGrid.grid.width * multiplier;
        const targetHeight = fracGrid.grid.height * multiplier;

        this.grid = new Grid(targetWidth, targetHeight, Color.BLACK);
        const maxX = targetWidth - 1;
        const maxY = targetHeight - 1;

        const transforms: Transform[] = (() => {
            switch (pattern) {
                case Pattern.Snowflake:
                    return [
                        { transpose: false, reflectX: false, reflectY: false },
                        { transpose: false, reflectX: true, reflectY: false },
                        { transpose: false, reflectX: false, reflectY: true },
                        { transpose: false, reflectX: true, reflectY: true },
                    ];
                case Pattern.Pinwheel:
                    return [
                        { transpose: false, reflectX: false, reflectY: false },
                        { transpose: true, reflectX: true, reflectY: false },
                        { transpose: true, reflectX: false, reflectY: true },
                        { transpose: false, reflectX: true, reflectY: true },
                    ];
                case Pattern.Fiducial:
                    return [
                        { transpose: false, reflectX: false, reflectY: false },
                    ];
            }
        })();

        const fracWidth = fracGrid.grid.width;
        const fracHeight = fracGrid.grid.height;
        for (let y = 0; y < fracHeight; y++) {
            for (let x = 0; x < fracWidth; x++) {
                const value = fracGrid.grid.getValue(x, y);
                const color = gradient(value);
                for (const t of transforms) {
                    let px = x;
                    let py = y;
                    if (t.transpose) {
                        [px, py] = [py, px];
                    }
                    if (t.reflectX) {
                        px = maxX - px;
                    }
                    if (t.reflectY) {
                        py = maxY - py;
                    }
                    this.grid.setValue(color, px, py);
                }
            }
        }
    }

    colors(): number[] {
        const result: number[] = new Array(this.grid.storage.length * 3);
        let idx = 0;
        for (const c of this.grid.storage) {
            result[idx++] = c.r;
            result[idx++] = c.g;
            result[idx++] = c.b;
        }
        return result;
    }
}

File: csharp/BCLifeHash/BCLifeHash/ColorGrid.cs
namespace BlockchainCommons.BCLifeHash;

internal sealed class ColorGrid
{
    public Grid<Color> Grid { get; }

    private readonly record struct Transform(bool Transpose, bool ReflectX, bool ReflectY);

    public ColorGrid(FracGrid fracGrid, Func<double, Color> gradient, Pattern pattern)
    {
        var multiplier = pattern == Pattern.Fiducial ? 1 : 2;
        var targetWidth = fracGrid.Grid.Width * multiplier;
        var targetHeight = fracGrid.Grid.Height * multiplier;

        Grid = new Grid<Color>(targetWidth, targetHeight);
        var maxX = targetWidth - 1;
        var maxY = targetHeight - 1;

        Transform[] transforms = pattern switch
        {
            Pattern.Snowflake =>
            [
                new(false, false, false),
                new(false, true, false),
                new(false, false, true),
                new(false, true, true),
            ],
            Pattern.Pinwheel =>
            [
                new(false, false, false),
                new(true, true, false),
                new(true, false, true),
                new(false, true, true),
            ],
            Pattern.Fiducial =>
            [
                new(false, false, false),
            ],
            _ => throw new InvalidOperationException("Unknown pattern"),
        };

        var fracWidth = fracGrid.Grid.Width;
        var fracHeight = fracGrid.Grid.Height;
        for (var y = 0; y < fracHeight; y++)
        {
            for (var x = 0; x < fracWidth; x++)
            {
                var value = fracGrid.Grid.GetValue(x, y);
                var color = gradient(value);
                foreach (var t in transforms)
                {
                    var px = x;
                    var py = y;
                    if (t.Transpose)
                        (px, py) = (py, px);
                    if (t.ReflectX)
                        px = maxX - px;
                    if (t.ReflectY)
                        py = maxY - py;
                    Grid.SetValue(color, px, py);
                }
            }
        }
    }

    public double[] Colors()
    {
        var result = new double[Grid.Storage.Length * 3];
        for (var i = 0; i < Grid.Storage.Length; i++)
        {
            var c = Grid.Storage[i];
            result[i * 3] = c.R;
            result[i * 3 + 1] = c.G;
            result[i * 3 + 2] = c.B;
        }
        return result;
    }
}
